### ISSUE TITLE
Implement Program.withCanvas

### ISSUE THREAD

**Issue by MangelMaxime** [MangelMaxime/Elmish.Canvas#2]
Implement `Program.withCanvas` for app that use only the canvas.

Example:
- Games
- Applications that needs control over their framerate

**Comment by alfonsogarciacaro** [comment]
I've been thinking about this. The current way is very nice to integrate canvas and elmish react, but it requires the React dependency and forces a virtual Dom recalculation for every frame.

It could be a good idea to think of the canvas-alone case first and then integration with React. This could mean to have a custom view and update functions, but it can also make the overall model cleaner. I'll try to PR a proof of concept later.

**Comment by MangelMaxime** [comment]
To be honest, I am facing a lot of performance problem with the `Segments Follow Mouse`.

Even by using a direct API (cf branch `test_stateful_canvas_direct_api`), the pressure on the GC is still here. I thinking removing React is a first step in the right direction. 

I also believe that the MailboxProcessor is also responsible for part of this pressure.

⚠️  The next part is to be taken with caution. ⚠️

**Comment by MangelMaxime** [comment]
To be honest, I am facing a lot of performance problem with the `Segments Follow Mouse`. And the performance doesn't seem to be constant between several application launches.

Even by using a direct API (cf branch `test_stateful_canvas_direct_api`), the pressure on the GC is still here. I thinking removing React is a first step in the right direction. 

I also believe that the MailboxProcessor is also responsible for part of this pressure.

⚠️  The next part is to be taken with caution. ⚠️ 

When looking at Elm application I don't see that much pressure on the GC... Also, if I remember correctly some tests I did in the past, even an empty/very small (like a time displayer) Elmish application seems to make the GC works.

I don't know well profiling, GC and things like that but I think there is something wrong somewhere.

**Comment by MangelMaxime** [comment]
About the start/stop signal, we could use a subscription that could trigger an event. Then, the program can listen to this event to start/stop the execution.

Why create an extra layer on top of the browser API? 

I prefer to pass the `ctx` to the user and let him use the native API directly:

```fs
    let draw (ctx : Browser.CanvasRenderingContext2D) (particle : Particle) (total : int) (width : float) =
        let angle = Math.Atan2(particle.Dy, particle.Dx)
        let scale = Math.Cos(Math.PI / 2. * (float particle.Id / float total))

        ctx.save()
        ctx.translate(particle.X, particle.Y)
        ctx.rotate(angle)
        ctx.scale(scale, scale)

        ctx.beginPath()
        ctx.moveTo(-width / 2. * 1.732, -width / 2.)
        ctx.lineTo(0. ,0.)
        ctx.lineTo(-width / 2. * 1.732, width / 2.)
        ctx.lineTo(-width / 2. * 1.2, 0.)
        ctx.fillStyle <- U3.Case1 "white"
        ctx.fill()

        ctx.restore()
```

We can of course later provide helpers function like `path`. For that, we have several options like provide a function or extends the `CanvasRenderingContext2D` class with new methods.

Having the code more imperative isn't a problem for me. All we are doing is still sending instructions to the browser.

About your implementation, I think we need to provide more information in the update function. In order to have smooth animations, there is a lot of things that the user need to know. [A Detailed Explanation of JavaScript Game Loops and Timing](https://isaacsukin.com/news/2015/01/detailed-explanation-javascript-game-loops-and-timing).

We can, of course, provide different `Program` signatures as Elmish do. With `Program.mkProgram` and `Program.mkSimpleProgram`.

Great thanks for your test, it gives me some ideas. I will try to implement my different points from it and share the result when ready :)

**Comment by alfonsogarciacaro** [comment]
No, I'm not extending `CanvasRenderingContext2D`. I thought about that but it could be a bit confusing for the user because autocomplete would show too many options. In any case, what I wrote was just a quick draft and, yes, the user should be able to use the native API directly :)

About start/stop, I think it needs to be flexible. For example, if we are using React to mount the canvas element, we need to send the signal to start from the `componentDidMount` method (and stop from `componentWillUnmount`). And we also may want to mix this with a pause button.

**Comment by MangelMaxime** [comment]
> About start/stop, I think it needs to be flexible. 

I agree. An event is pretty flexible because you can control its scope and if you have access to its scope you can do whatever you want.

About the react, mapping I don't have it in mind for now because I prefer to solve one problem first. Having a nice user experience for canvas usage with Fable. And later we can think about react integration. I am not much worried about this one because you can simply create a stateful component if we want in order to minimize the impact on an Elmish application.
